cpp/src/join/hash_join.cuh

```diff
@@ -0,0 +1,336 @@
+  // Even if the hash table will be empty, it still must be allocated for the
+  // probing phase in the event of an outer join
+  size_t const hash_table_size = compute_hash_table_size(build_table_num_rows);
+
+  auto hash_table = multimap_type::create(
+      hash_table_size, true,
+      multimap_type::hasher(),
+      multimap_type::key_equal(),
+      multimap_type::allocator_type(),
+      stream
+      );
+
+  // build the hash table
+  if (build_table_num_rows > 0) {
+    row_hash hash_build{*build_table};
+    rmm::device_scalar<int> failure(0, stream);
+    constexpr int block_size{DEFAULT_JOIN_BLOCK_SIZE};
+    experimental::detail::grid_1d config(build_table_num_rows, block_size);
+    build_hash_table<<<config.num_blocks, config.num_threads_per_block, 0, stream>>>(
+        *hash_table,
+        *build_table,
+        hash_build,
+        build_table_num_rows,
+        failure.data());
+    // Check error code from the kernel
+    if (failure.value() == 1) { CUDF_FAIL("Hash Table insert failure."); }
+  }
+
+  auto estimated_join_output_size =
+    estimate_join_output_size<JoinKind, multimap_type>(
+        *build_table, *probe_table, *hash_table, stream);
+
+  size_type estimated_size = estimated_join_output_size->value(stream);
+  // If the estimated output size is zero, return immediately
+  if (estimated_size == 0) {
+    rmm::device_vector<size_type> left_empty, right_empty;
+    return std::make_pair(left_empty, right_empty);
+  }
+
+  // Because we are approximating the number of joined elements, our approximation
+  // might be incorrect and we might have underestimated the number of joined elements.
+  // As such we will need to de-allocate memory and re-allocate memory to ensure
+  // that the final output is correct.
+  rmm::device_scalar<size_type> write_index(0, stream);
+  size_type join_size{0};
+
+  rmm::device_vector<size_type> left_indices;
+  rmm::device_vector<size_type> right_indices;
+  do {
+    left_indices.resize(estimated_size);
+    right_indices.resize(estimated_size);
+
+    constexpr int block_size{DEFAULT_JOIN_BLOCK_SIZE};
+    experimental::detail::grid_1d config(probe_table->num_rows(), block_size);
+    write_index.set_value(0);
+
+    row_hash hash_probe{*probe_table};
+    row_equality equality{*probe_table, *build_table};
+    probe_hash_table<JoinKind,
+                     multimap_type,
+                     hash_value_type,
+                     block_size,
+                     DEFAULT_JOIN_CACHE_SIZE>
+    <<<config.num_blocks, config.num_threads_per_block, 0, stream>>>(
+        *hash_table,
+        *build_table,
+        *probe_table,
+        hash_probe,
+        equality,
+        probe_table->num_rows(),
+        left_indices.data().get(),
+        right_indices.data().get(),
+        write_index.data(),
+        estimated_size,
+        flip_join_indices);
+
+    CUDA_CHECK_LAST();
+
+    join_size = write_index.value();
+    if (estimated_size < join_size) {
```

Comment:
Now that you switched to do..while I think you can remove this if now and just always double the estimated_size at the end of the loop (since the loop will exit when it's bigger than join_size).